File: Medium/0005.LongestPalindromicSubstring/longestPalSubStr.java
public class Solution {
    public String longestPalindrome(String s) {
        if (s == null || s.length() < 1) return "";
        int start = 0, end = 0;
        
        // Itera su ogni carattere come potenziale centro
        for (int i = 0; i < s.length(); i++) {
            // Palindromi dispari (centro in i)
            int[] odd = expand(s, i, i);
            // Palindromi pari (centro tra i e i+1)
            int[] even = expand(s, i, i + 1);
            
            // Aggiorna gli indici del palindromo più lungo
            if (odd[1] - odd[0] > end - start) {
                start = odd[0];
                end = odd[1];
            }
            if (even[1] - even[0] > end - start) {
                start = even[0];
                end = even[1];
            }
        }
        // Estrae la sottostringa (end+1 perché substring esclude l'indice finale)
        return s.substring(start, end + 1);
    }
    
    // Funzione helper per espandere intorno a un centro
    private int[] expand(String s, int l, int r) {
        while (l >= 0 && r < s.length() && s.charAt(l) == s.charAt(r)) {
            l--;
            r++;
        }
        // Regola gli indici all'ultimo palindromo valido
        return new int[]{l + 1, r - 1};
    }
}

File: Medium/0005.LongestPalindromicSubstring/longestPalSubStr.go
package main

func longestPalindrome(s string) string {
	if len(s) < 1 {
		return ""
	}

	start, end := 0, 0
	// Itera su ogni carattere come potenziale centro
	for i := 0; i < len(s); i++ {
		// Palindromi dispari (centro in i)
		l1, r1 := expand(s, i, i)
		// Palindromi pari (centro tra i e i+1)
		l2, r2 := expand(s, i, i+1)

		// Aggiorna gli indici del palindromo più lungo
		if r1-l1 > end-start {
			start, end = l1, r1
		}
		if r2-l2 > end-start {
			start, end = l2, r2
		}
	}
	return s[start : end+1] // Estrae la sottostringa
}

// Funzione helper per espandere intorno a un centro
func expand(s string, l int, r int) (int, int) {
	for l >= 0 && r < len(s) && s[l] == s[r] {
		l--
		r++
	}
	// Regola gli indici all'ultimo palindromo valido
	return l + 1, r - 1
}
